<link href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" rel="stylesheet">

#### Welcome to CEVIChE (CEll VIability Calculator from gene Expression)
This app allows to browse matched and predicted cell viaiblity values for the LINCS-L1000 perturbation dataset, and also to predict cell viability for your own gene expression data.

<i class="far fa-comment"></i> designed to interactively explore the results of <a href="https://doi.org/10.1093/nar/gkz805" target="_blank">*Signatures of cell death and proliferation in perturbation transcriptomics data - from confounding factor to effective prediction*</a>.

<i class="fas fa-database"></i> database and model created by Bence Szalai (<a href="mailto:szalai.bence@med.semmelweis-univ.hu" target="_blank"><i class="glyphicon glyphicon-envelope"></i></a>) at the <a href="http://semmelweis.hu/elettan/en/" target="_blank">Department of Physiology</a>, Faculty of Medicine, Semmelwies University.


<i class="fas fa-user-cog"></i> application developed and maintained by Christian Holland (<a href="mailto:christian.holland@bioquant.uni-heidelberg.de" target="_blank"><i class="glyphicon glyphicon-envelope"></i></a>) at the <a href="http://saezlab.org" target="_blank">Saezlab</a>, Institute for Computational Biomedicine, University of Heidelberg.

<i class="fas fa-laptop-code"></i> developed using <a href="https://shiny.rstudio.com" target="_blank">Shiny</a>.

---

<i class="fab fa-github"></i> Source code is available on <a href="https://github.com/bence-szalai/Cell-death-signatures" target="_blank">GitHub</a>.

<i class="fas fa-question"></i> Please use <a href="https://github.com/bence-szalai/Cell-death-signatures/issues" target="_blank">GitHub's issue system</a> to address questions, bugs or feedback. 

---

<center>
<a href="http://saezlab.org" target="_blank"><img src="../www/logo_saezlab.png" width="20%", align="center"></a>

<i class="fab fa-twitter"></i> <a href="https://twitter.com/sysbiomed?lang=de" target="_blank">@sysbiomed</a>
</center>
